File: clients/Go/model/build_properties.go
package model

import "io"
import . "aicup2020/stream"

type BuildProperties struct {
    Options []EntityType
    InitHealth *int32
}
func NewBuildProperties(options []EntityType, initHealth *int32) BuildProperties {
    return BuildProperties {
        Options: options,
        InitHealth: initHealth,
    }
}
func ReadBuildProperties(reader io.Reader) BuildProperties {
    result := BuildProperties {}
    result.Options = make([]EntityType, ReadInt32(reader))
    for i := range result.Options {
        result.Options[i] = ReadEntityType(reader)
    }
    if ReadBool(reader) {
        var InitHealthValue int32
        InitHealthValue = ReadInt32(reader)
        result.InitHealth = &InitHealthValue
    } else {
        result.InitHealth = nil
    }
    return result
}
func (value BuildProperties) Write(writer io.Writer) {
    WriteInt32(writer, int32(len(value.Options)))
    for _, OptionsElement := range value.Options {
        WriteInt32(writer, int32(OptionsElement))
    }
    if value.InitHealth == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        WriteInt32(writer, (*value.InitHealth))
    }
}

File: clients/Python-PyPy/model/build_properties.py
from .entity_type import EntityType
class BuildProperties:
    def __init__(self, options, init_health):
        self.options = options
        self.init_health = init_health
    @staticmethod
    def read_from(stream):
        options = []
        for _ in range(stream.read_int()):
            options_element = EntityType(stream.read_int())
            options.append(options_element)
        if stream.read_bool():
            init_health = stream.read_int()
        else:
            init_health = None
        return BuildProperties(options, init_health)
    def write_to(self, stream):
        stream.write_int(len(self.options))
        for element in self.options:
            stream.write_int(element)
        if self.init_health is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            stream.write_int(self.init_health)
    def __repr__(self):
        return "BuildProperties(" + \
            repr(self.options) + "," + \
            repr(self.init_health) + \
            ")"
